File: python/05/star_one.py
import os

input_file = os.path.join(os.path.dirname(__file__), "../../input/day_05.txt")

with open(input_file, "r") as file:
    input = file.read()

seeds, *maps = input.split("\n\n")
seeds_arr = [int(x) for x in seeds.split(": ")[1].split(" ")]

seeds_dct = {x: x for x in seeds_arr}
for x in maps:
    for y in x.split("\n")[1:]:
        dst_start, src_start, rng = [int(z) for z in y.split(" ")]

        for seed in seeds_dct.keys():
            if seed >= src_start and seed <= src_start + rng - 1:
                seeds_dct[seed] = dst_start + (seed - src_start)

    seeds_dct = {x: x for x in seeds_dct.values()}
print(min(seeds_dct.values()))

File: typescript/05/star_one.ts
import * as fs from "fs/promises";
import * as path from "path";

(async () => {
    const rawInput = await fs.readFile(path.resolve(__dirname, '../../input/day_05.txt'), "utf-8");
    const arr = rawInput.split("\n\n");

    const [seeds, ...maps] = arr;
    const seedsArr = seeds.split(": ")[1].split(" ").map(Number)
    let seedsDict: {[key: number]: number} = seedsArr.reduce((acc, val) => ({ ...acc, [val]: val }), {});

    maps.forEach(x => {
        x.split("\n").slice(1).forEach(y => {
            const [dstStart, srcStart, rng] = y.split(" ").map(Number)

            Object.keys(seedsDict).forEach(seed => {
                const seedInt = Number(seed)  // behind the scenes, keys are parsed as strings
                if (seedInt >= srcStart && seedInt <= srcStart + rng - 1)
                    seedsDict[seed] = dstStart + (seedInt - srcStart)
            })
        })

        seedsDict = Array.from(Object.values(seedsDict)).reduce((acc, val) => ({ ...acc, [val]: val }), {})
    })
    const total = Math.min(...Object.values(seedsDict))
    console.log(total)
})();